# 03 - Checking Data Consistency

The purpose of this document is to ensure there is **no Noisy Data** in the interim dataset as the output of the *ETL.py* script.

## Setting Up Project Directory

In [24]:
from jupyter_init import setup

setup()

from src_code.config import *

## Loading Dataset

In [25]:
import pandas as pd
import numpy as np

DF_PATH = ETL_MAPPINGS['test']['current_newest']

# ---- LOAD ----
df = pd.read_feather(DF_PATH)
print(f"Loaded dataframe with {len(df)} rows and {len(df.columns)} columns\n")

df.dtypes

Loaded dataframe with 7363 rows and 31 columns



datetime                      datetime64[us, pytz.FixedOffset(-60)]
commit                                                       object
repo                                                         object
filepath                                                     object
content                                                      object
methods                                                      object
lines                                                        object
author_email                                                 object
canonical_datetime                              datetime64[ns, UTC]
author_exp_pre                                                int64
author_recent_activity_pre                                    int64
label                                                         int64
loc_added                                                     int64
loc_deleted                                                   int64
files_changed                                   

## Converting NumpyArray -> List



In [26]:
#Convert the NumPy arrays back to Python lists
for col in ['code_embed', 'msg_embed', 'methods', 'lines']:
    # Use .apply(list) or .apply(lambda x: x.tolist()) for robustness
    df[col] = df[col].apply(list)

# print(df['content'].head(5))

## Missing Value Audit

In [27]:
print("## 1. Missing Values per Column")
nulls = df.isnull().sum().sort_values(ascending=False)
print(nulls.to_markdown())

## 1. Missing Values per Column
|                            |   0 |
|:---------------------------|----:|
| datetime                   |   0 |
| commit                     |   0 |
| repo                       |   0 |
| filepath                   |   0 |
| content                    |   0 |
| methods                    |   0 |
| lines                      |   0 |
| author_email               |   0 |
| canonical_datetime         |   0 |
| author_exp_pre             |   0 |
| author_recent_activity_pre |   0 |
| label                      |   0 |
| loc_added                  |   0 |
| loc_deleted                |   0 |
| files_changed              |   0 |
| hunks_count                |   0 |
| msg_len                    |   0 |
| has_fix_kw                 |   0 |
| has_bug_kw                 |   0 |
| ast_delta                  |   0 |
| complexity_delta           |   0 |
| max_func_change            |   0 |
| time_since_last_change     |   0 |
| todo                       |   0 |
| fixm

## Primary Key Integrity

In [28]:
print("## 2. Primary Key Uniqueness Check")
key_cols = ["repo", "commit", "filepath"]

dupes = df.duplicated(subset=key_cols).sum()
print(f"Duplicate key rows: {dupes}")

## 2. Primary Key Uniqueness Check
Duplicate key rows: 0


## Label Distribution

In [29]:
print("## 3. Label Distribution")
print(df['label'].value_counts(normalize=True).to_markdown())

## 3. Label Distribution
|   label |   proportion |
|--------:|-------------:|
|       0 |     0.887546 |
|       1 |     0.112454 |


## Repository Distribution (Imbalance Check)

In [30]:
print("## 4. Repository Distribution")
repo_dist = df['repo'].value_counts(normalize=True)
print(repo_dist.to_markdown())

## 4. Repository Distribution
| repo    |   proportion |
|:--------|-------------:|
| sentry  |    0.317941  |
| ray     |    0.257911  |
| core    |    0.237811  |
| airflow |    0.0885509 |
| pandas  |    0.0670922 |
| ansible |    0.030694  |


## Value Range Scan for Numeric Columns

Automatically detects:

- negatives where not allowed
- max values
- suspicious spikes

In [31]:
print("## 5. Numeric Column Range Scan")
num_cols = df.select_dtypes(include=[np.number]).columns

ranges = pd.DataFrame({
    "min": df[num_cols].min(),
    "median": df[num_cols].median(),
    "mean": df[num_cols].mean(),
    "max": df[num_cols].max()
})

print(ranges.to_markdown())

## 5. Numeric Column Range Scan
|                            |   min |   median |          mean |    max |
|:---------------------------|------:|---------:|--------------:|-------:|
| author_exp_pre             |     0 |       19 |   40.0522     |    394 |
| author_recent_activity_pre |     0 |       12 |   24.9315     |    222 |
| label                      |     0 |        0 |    0.112454   |      1 |
| loc_added                  |     0 |       31 |  109.042      |   1955 |
| loc_deleted                |     0 |       58 |  139.087      |   1985 |
| files_changed              |     0 |        6 |   18.3603     |    136 |
| hunks_count                |     0 |       32 |   85.9383     |    826 |
| msg_len                    |    16 |      128 |  229.56       |   8064 |
| has_fix_kw                 |     0 |        0 |    0.21513    |      1 |
| has_bug_kw                 |     0 |        0 |    0.0309656  |      1 |
| ast_delta                  |     0 |       21 |  172.508      |  1

## Check Columns Expected to Be Non-Negative

In [32]:
non_negative_cols = [
    "loc_added", "loc_deleted",
    "files_changed", "hunks_count",
    "msg_len", "ast_delta",
    "complexity_delta", "max_func_change",
    "author_exp_pre", "author_recent_activity_pre",
    "todo", "fixme", "try", "except", "raise",
    "recent_churn"
]

print("## 6. Negative Value Check")
for col in non_negative_cols:
    bad = (df[col] < 0).sum()
    print(f"{col}: {bad} negative values")

## 6. Negative Value Check
loc_added: 0 negative values
loc_deleted: 0 negative values
files_changed: 0 negative values
hunks_count: 0 negative values
msg_len: 0 negative values
ast_delta: 0 negative values
complexity_delta: 0 negative values
max_func_change: 0 negative values
author_exp_pre: 0 negative values
author_recent_activity_pre: 0 negative values
todo: 0 negative values
fixme: 0 negative values
try: 0 negative values
except: 0 negative values
raise: 0 negative values
recent_churn: 0 negative values


## Suspicious Feature Check: time_since_last_change

In [33]:
print("## 7. time_since_last_change Outliers")
tslc = df["time_since_last_change"]

print(f"Negative values: {(tslc < 0).sum()}")
print(f"99.9% quantile: {tslc.quantile(0.999)}")
print(f"Min: {tslc.min()}")
print(f"Max: {tslc.max()}")

## 7. time_since_last_change Outliers
Negative values: 0
99.9% quantile: 346372.0
Min: 0
Max: 659000


### Understanding the Feature

*time_since_last_change = c.committed_date - last_time*

Where:
- c.committed_date = current commit timestamp (UNIX seconds)
- last_time = timestamp of first parent commit

So the feature = time difference between consecutive commits.

This represents how much time passed between commits in a repo.

### Why Negative Values?

Meaning:

- Some commits appear to be ~4.6 days negative (-396,818 sec)
- Some commits appear to be ~77 days ahead (6.6M sec)

This is expected when real Git data is used.

Git timestamps can go backwards because:

#### (1) Rebased or rewritten history

During rebases, old commits appear “later” than newer ones.

*WHY?*

When you merge/rebase, Git does not reorder commits chronologically.
Instead, it preserves the logical order of development.

#### (2) Merge parents

You use only the first parent:

if c.parents:
    last_time = c.parents[0].committed_date


But merges may introduce non-linear time ordering.

#### (3) Clock drift

Different authors → different local machine clocks.

#### (4) Shallow clones or incomplete history

If the repo is shallow-fetched, parent commits may have weird timestamps.

None of this indicates your extraction is wrong.



## Binary Flag Integity

In [34]:
print("## 8. Binary Columns Integrity")
bin_cols = ["has_fix_kw", "has_bug_kw"]

for col in bin_cols:
    bad = df[~df[col].isin([0,1])]
    print(f"{col}: {len(bad)} invalid values")

## 8. Binary Columns Integrity
has_fix_kw: 0 invalid values
has_bug_kw: 0 invalid values


## Embedding Consistency Check

Ensure:

- no None
- all lists
- identical dimensionality

In [35]:
# # Convert the NumPy arrays back to Python lists
# for col in ['code_embed', 'msg_embed']:
#     # Use .apply(list) or .apply(lambda x: x.tolist()) for robustness
#     df[col] = df[col].apply(list)

print(type(df.loc[0, 'code_embed']))
print(type(df.loc[0, 'msg_embed']))

print("## 9. Embedding Structural Checks")

# None count
print("code_embed None count:", df['code_embed'].isna().sum())
print("msg_embed None count:", df['msg_embed'].isna().sum())

# Check if all are lists
print("\nNon-list code_embed rows:", (~df['code_embed'].apply(lambda x: isinstance(x, list))).sum())
print("Non-list msg_embed rows:", (~df['msg_embed'].apply(lambda x: isinstance(x, list))).sum())

# Check dimensionality
dims = df['code_embed'].apply(lambda x: len(x) if isinstance(x, list) else None)
print("\nEmbedding dimensionality distribution:")
print(dims.value_counts().head())

<class 'list'>
<class 'list'>
## 9. Embedding Structural Checks
code_embed None count: 0
msg_embed None count: 0

Non-list code_embed rows: 0
Non-list msg_embed rows: 0

Embedding dimensionality distribution:
code_embed
768    7363
Name: count, dtype: int64


## Datetime Consistency

Check for:

- NaT values
- ordering sanity (commit should not be older than file's previous record)

In [36]:
print("## 10. Datetime Columns Audit")

date_cols = ["datetime", "canonical_datetime"]

for col in date_cols:
    print(f"{col}: NaT count = {df[col].isna().sum()}")
    print(f"{col}: min = {df[col].min()}, max = {df[col].max()}")

## 10. Datetime Columns Audit
datetime: NaT count = 0
datetime: min = 2022-06-15 08:05:34-01:00, max = 2022-11-03 19:53:36-01:00
canonical_datetime: NaT count = 0
canonical_datetime: min = 2022-06-15 08:45:47+00:00, max = 2022-11-03 00:00:15+00:00


In [37]:
print("## 12. Columns With Only One Unique Value")

for col in df.columns:
    # Convert arrays to tuples for uniqueness check
    if df[col].apply(lambda x: isinstance(x, (np.ndarray, list))).any():
        # Convert each element to a tuple (or a string representation)
        # print(col)
        unique_values = df[col].apply(lambda x: tuple(x) if isinstance(x, (np.ndarray, list)) else x)
    else:
        unique_values = df[col]

    if unique_values.nunique(dropna=True) == 1:
        print(f"⚠️ {col} has only one unique value")

## 12. Columns With Only One Unique Value


## Check Text Columns for Weirdness

Empty strings? Too short? Too long?

In [38]:
print("## 13. Text Field Checks")

if 'content' in df.columns:
    print("Empty content rows:", (df['content'].str.len() == 0).sum())
    print(df['content'].str.len().describe())

# Check 'methods' if it is a list column
if 'methods' in df.columns:
    # Use len() on the Python lists
    print("\nEmpty methods rows:", (df['methods'].apply(len) == 0).sum())
    print(df['methods'].apply(len).describe())

## 13. Text Field Checks
Empty content rows: 30
count      7363.000000
mean       1697.582371
std        7573.354188
min           0.000000
25%         476.000000
50%         835.000000
75%        1766.000000
max      618763.000000
Name: content, dtype: float64

Empty methods rows: 2077
count    7363.000000
mean        1.991987
std         3.684710
min         0.000000
25%         0.000000
50%         1.000000
75%         2.000000
max       118.000000
Name: methods, dtype: float64


## Check For Impossible Values

In [39]:
print("## 14. Logical Consistency Checks")

# msg_len should match commit message length
if "msg_len" in df.columns:
    print("msg_len outliers (msg_len <= 0):", (df['msg_len'] <= 0).sum())

## 14. Logical Consistency Checks
msg_len outliers (msg_len <= 0): 0


## Filepath Sanity
Check for Windows vs POSIX weirdness.

In [40]:
# Normalize all filepaths to use forward slashes
df['filepath'] = df['filepath'].str.replace('\\', '/', regex=False)

# Check again for paths without a slash
bad_paths = df[~df['filepath'].str.contains('/')]
print("Filepaths without / (unexpected after normalization):", len(bad_paths))
print(df.loc[~df['filepath'].str.contains('/'), 'filepath'].head(20))

Filepaths without / (unexpected after normalization): 13
23           setup.py
67           setup.py
156          setup.py
481          setup.py
2228         setup.py
2958         setup.py
3014    versioneer.py
5093         setup.py
5434         setup.py
5863         setup.py
6103         setup.py
6307         setup.py
6308         setup.py
Name: filepath, dtype: object


## Check Recent Churn for Extreme Outliers

In [41]:
print("## 16. recent_churn Outlier Scan")
print(df['recent_churn'].describe())
print("99.9% quantile:", df['recent_churn'].quantile(0.999))

## 16. recent_churn Outlier Scan
count      7363.000000
mean       6575.006519
std       16244.949940
min           0.000000
25%         100.000000
50%         822.000000
75%        3891.000000
max      135911.000000
Name: recent_churn, dtype: float64
99.9% quantile: 127555.13000000094


## Check Distribution of Code Activity Keywords

(todo, fixme, try/except/raise)

In [42]:
print("## 17. Keyword Column Distributions")
kw_cols = ["todo", "fixme", "try", "except", "raise"]

print(df[kw_cols].describe().T.to_markdown())

## 17. Keyword Column Distributions
|        |   count |       mean |        std |   min |   25% |   50% |   75% |   max |
|:-------|--------:|-----------:|-----------:|------:|------:|------:|------:|------:|
| todo   |    7363 | 0.186609   |  1.07322   |     0 |     0 |     0 |     0 |    17 |
| fixme  |    7363 | 0.00203721 |  0.0480105 |     0 |     0 |     0 |     0 |     2 |
| try    |    7363 | 4.34931    | 26.5048    |     0 |     0 |     0 |     2 |  1422 |
| except |    7363 | 1.67513    | 10.9471    |     0 |     0 |     0 |     0 |   256 |
| raise  |    7363 | 0.711395   |  3.47064   |     0 |     0 |     0 |     0 |    54 |


## Duplicate commit-message / code-embed lengths

Check uniformity:

In [43]:
code_duplicates = df['code_embed'].apply(len).value_counts().head()
msg_duplicates = df['msg_embed'].apply(len).value_counts().head()

print(code_duplicates)
print(msg_duplicates)

code_embed
768    7363
Name: count, dtype: int64
msg_embed
768    7363
Name: count, dtype: int64


So all your embeddings being length 768 means:

- The model you are using outputs a 768-dimensional vector
- The embedding extraction process worked for every commit
- No corrupted or empty embeddings
- No input missing (no None, no NaN, no empty list, etc.)

## File extension distribution
This reveals if:

- non-Python files contaminate AST/keyword metrics
- certain extensions dominate bugs

In [44]:
df['ext'] = df['filepath'].str.split('.').str[-1]
df['ext'].value_counts()

ext
py     7355
pyi       8
Name: count, dtype: int64

This does not mean the entire Pandas repo only contains .py files.
It means:

All diffs included in your dataset modify .py or .pyi files.

The Defectors dataset:

- extracts commits that add/remove lines regarded as “faulty”
- focuses on logical/code changes, not text, docs, or build files
- filters out non-source changes for consistency

So your dataset is a filtered view of the repo, not the repo itself.

## Diff size sanity

Large mismatches could indicate:

- truncated diffs
- non-standard diff formatting
- metadata lines (prefix +++, ---, etc.)

In [45]:
(df["content"].str.count("\n") - df["loc_added"] - df["loc_deleted"]).describe()


count     7363.000000
mean      -204.758115
std        433.672633
min      -3627.000000
25%       -257.000000
50%        -65.000000
75%         -2.000000
max      18363.000000
dtype: float64

The discrepancy between content line counts and loc_added + loc_deleted is normal in real Git data.

Reasons:

1. Multi-line statements / code folding in diffs
2. Partial line changes counted in hunks
3. Files with removed lines only (content now shorter)
4. Large diffs in a single commit skew the stats

No indication of extraction errors — these are just properties of real commit diffs.